batch/batch/server/server.py

```diff
@@ -158,62 +151,112 @@ def _has_next_task(self):
+                name='gsa-key')]
+        volume_mounts = [
+            kube.client.V1VolumeMount(
+                mount_path='/gsa-key',
+                name='gsa-key')]
+
+        if len(self._tasks) > 1:
+            if self._pvc_name is None:
+                self._pvc_name = await self._create_pvc()
+                if self._pvc_name is None:
+                    log.info(f'could not create pod for job {self.id} due to pvc creation failure')
+                    return
+            volumes.append(kube.client.V1Volume(
+                persistent_volume_claim=kube.client.V1PersistentVolumeClaimVolumeSource(
+                    claim_name=self._pvc_name),
+                name=self._pvc_name))
+            volume_mounts.append(kube.client.V1VolumeMount(
+                mount_path='/io',
+                name=self._pvc_name))
+
+        pod_spec = v1.api_client._ApiClient__deserialize(self._current_task.pod_spec_dict, kube.client.V1PodSpec)
+        if pod_spec.volumes is None:
+            pod_spec.volumes = []
+        pod_spec.volumes.extend(volumes)
+        for container in pod_spec.containers:
+            if container.volume_mounts is None:
+                container.volume_mounts = []
+            container.volume_mounts.extend(volume_mounts)
+
+        pod_template = kube.client.V1Pod(
+            metadata=kube.client.V1ObjectMeta(
+                generate_name='job-{}-{}-'.format(self.id, self._current_task.name),
+                labels={'app': 'batch-job',
+                        'hail.is/batch-instance': INSTANCE_ID,
+                        'uuid': uuid.uuid4().hex
+                        }),
+            spec=pod_spec)
 
-        pod = v1.create_namespaced_pod(
-            HAIL_POD_NAMESPACE,
-            self._current_task.pod_template,
-            _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
-        self._pod_name = pod.metadata.name
+        try:
+            pod = v1.create_namespaced_pod(
+                HAIL_POD_NAMESPACE,
+                pod_template,
+                _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+            self._pod_name = pod.metadata.name
 
-        await db.jobs.update_record(self.id,
-                                    pod_name=self._pod_name)
+            await db.jobs.update_record(self.id,
+                                        pod_name=self._pod_name)
 
-        log.info('created pod name: {} for job {}, task {}'.format(self._pod_name,
-                                                                   self.id,
-                                                                   self._current_task.name))
+            log.info('created pod name: {} for job {}, task {}'.format(self._pod_name,
+                                                                       self.id,
+                                                                       self._current_task.name))
+        except kube.client.rest.ApiException as err:
+            await self.mark_complete(None)
```

Comment:
Sorry, I should have said this shouldn't be marked complete.  What if the create_pod call failed because there was a temporary network issue?  We should definitely retry.  However, if there is, say, the pod spec is invalid, we consider it failed.  However, that should be listed explicitly, and unknown errors should be retries.  So I think doing nothing here is correct, and the next iteration of the refresh loop will try to recreate the pod.